Left shift in StaticVector::erase(), add tests

Yggdrasill opened the pull request:
Hello,

While I was working on using `StaticVector` for vendor item arrays, I noticed that `StaticVector::erase()` doesn't seem to behave as expected. I asked about this on Discord and @glebm concluded it is a bug, and says that it is intended to work exactly like `std::vector`'s erase. 

The following changes have been made to `static_vector.hpp`:

- `std::move` is called before `std::destroy_at`, shifting items left.
- `std::destroy_at()` now destroys the entries at the back of the array after the move.
- `StaticVector::clear()` returns early if there are no member items.

I am not an expert in C++ and am not 100% confident in this PR or the changes made, so feedback is greatly appreciated. 

Regards,
Yggdrasill

IAmTrial commented on Source/utils/static_vector.hpp:
```diff
@@ -71,12 +71,14 @@ class StaticVector {
 	T &operator[](std::size_t pos) { return *data_[pos].ptr(); }
 
-	void erase(const T *begin, const T *end)
+	void erase(T *begin, T *end)
```
The erase function **should** be accepting const iterator type. See the documentation for [`vector::erase`](https://en.cppreference.com/w/cpp/container/vector/erase).

Yggdrasill replied:
Given that shuffling the members left requires copying, how do you propose I can do this with const iterators? I am not sure how the C++ STL implements that in `std::vector`. Can you explain a little bit more, please?

IAmTrial replied:
There are two ways about it:
1. You are working with a non-const function and are intending to modify elements. Therefore, you can safely `const_cast` away the constness of the two inputs. C++ standard only indicates that a modification to objects that are truly const to be undefined behavior. The input iterators are not truly const, so therefore, modification has defined behavior.
2. If you are apprehensive about `const_cast`ing away the constness of the inputs, then you can use pointer arithmetic to determine the indexes of the two iterators, then apply that as the offset to `ptr()`. This will get you the original inputs as non-const, and without `const_cast`.

Yggdrasill replied:
Alright, I've done what you suggested, specifically option #2. Thanks!

chasedjones88 commented:
Consider using the term "Shift" instead of "Shuffle" as shuffle indicates, at least to me, an element of random reordering.

ephphatha commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+		expected.push_back(i);
+	}
+
+	for (int idx : erase_idx) {
+		/* This is actually to guard against gcc's vector implementation causing
+		 * crashes, rather than StaticVector
+		 */
+		if (expected.size() == 0) {
+			break;
+		}
+		container.erase(container.begin() + idx, container.begin() + idx + 1);
+		expected.erase(expected.begin() + idx);
+	}
```
Deleting an element from these containers changes the size of the containers, meaning all indexes greater than the deleted location no longer point to the originally intended location. If you want to handle random deletions you'll need to ensure you either delete from the end first (e.g. reverse sort the indexes to delete) or adjust your indexes to ensure they point to valid locations (e.g. decrement every index > the one you just deleted.

ephphatha commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+		expected.erase(expected.begin() + idx);
+	}
+
+	EXPECT_EQ(container.size(), expected.size());
+	for (size_t i = 0; i < expected.size(); i++) {
+		EXPECT_EQ(container[i], expected[i]);
+	}
```
if the random roll for `erasures` returns `size() - 1` or `size() - 2` (and assuming all deletions actually get processed) the containers end up empty (because you're always attempting to delete 2+erasures elements). That's a useful test however you'd want that as a separate test case or an explicit check after deleting some (but not all) of the contents.

ephphatha commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+	container.clear();
+	expected.clear();
+
+	EXPECT_EQ(container.size(), expected.size());
```
`container.size()` should be 0 after a clear. There's no point comparing it to another container.

ephphatha commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+		if (expected.size() == 0) {
+			break;
+		}
+		container.erase(container.begin() + idx, container.begin() + idx + 1);
```
This is a great argument for adding a `StaticVector::erase(T *element)` overload instead of requiring callers compute the end pointer themselves.

Yggdrasill replied:
I agree with this and was thinking about it too, as it more closely resembles `std::vector`'s erase.

glebm commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+using namespace devilution;
+
+namespace {
+StaticVector<int, MAX_SIZE> container;
```
Yeah just put this static vector in every test case

Yggdrasill commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+	std::vector<int> expected;
+
+	erase_idx.push_back(0);
+	erase_idx.push_back(container.size() - 1);
```
This is a good point and I greatly appreciate it. I think it made sense to separate out static test scenarios entirely into its own test routine, so that's what I ended up doing.

Yggdrasill commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,82 @@
+TEST(StaticVector, StaticVector_push_back)
+{
+	std::srand(clock());
+	size_t size = 10 + (std::rand() % (MAX_SIZE - 10));
```
Sorry about that, I'm not well-known for writing tests myself, and so haven't used GTest. Thank you!

Yggdrasill commented:
Sorry for the messy commit history. I always forget to run `clang-format` before pushing, and MSVC produced some test failures that I can't produce locally on Debian. I set up a VM to run these tests locally now.

There is at this point a possible change that I'd like to discuss here. Currently `StaticVector::erase()` returns early if erasures are requested out of bounds, and leaves the array untouched. What I would like to ask is if it's desired that `erase()` should clamp these requests to the bounds of the array, and operate normally. When called with either:

`container.erase(container.begin() - 1);`
`container.erase(container.end() - 1, container.end() + 1);`

Would it be better to exit early as it does presently, or should it fix up these values to within bounds?

AJenbo commented:
I think we both made the mistake thinking this would be a quick little change :D

ephphatha commented:
> Would it be better to return early as it does presently, or should it fix up these values to within bounds? Or should it perhaps do an assertion for these cases?

Adding asserts will be fine, I don't know exactly where in the spec it's defined but usually the expectation is the caller will do the right thing and not try erase invalid values.

For example, you can `assert(element >= begin() && element < end())` inside `erase(T *element)` to match the expected behaviour of std::vector.

> `container.erase(container.begin() - 1);`

would then trigger an assert failure (or crash in release builds).

and for `erase(T *first, T* last)` `assert(first >= begin() && last <= end() && first <= last)` is sufficient.

> `container.erase(container.end() - 1, container.end() + 1);`

would fail the assert

`container.erase(container.begin(), container.begin())` would be fine and result in a no-op, even if the container is empty.

Yggdrasill commented:
I've done that now, and also added tests for these assertions, though one thing I find quite confusing is that locally these tests pass just fine. Does anyone know why the tests are failing here? Specifically in the job logs:

```Note: Google Test filter = StaticVector.StaticVector_push_back_full
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from StaticVector
[ RUN      ] StaticVector.StaticVector_push_back_full
test/staticvector_test.cpp:44: Failure
Death test: container.push_back(size + 1)
    Result: failed to die.
 Error msg:
[  DEATH   ] 
[  FAILED  ] StaticVector.StaticVector_push_back_full (1 ms)
[----------] 1 test from StaticVector (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] StaticVector.StaticVector_push_back_full

 1 FAILED TEST

        Start 194: StaticVector.StaticVector_erase
194/272 Test #194: StaticVector.StaticVector_erase ......................................................................................................................................................................***Failed    0.03 sec
Note: Google Test filter = StaticVector.StaticVector_erase
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from StaticVector
[ RUN      ] StaticVector.StaticVector_erase
test/staticvector_test.cpp:54: Failure
Death test: container.erase(container.begin())
    Result: failed to die.
 Error msg:
[  DEATH   ] 
[  FAILED  ] StaticVector.StaticVector_erase (1 ms)
[----------] 1 test from StaticVector (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] StaticVector.StaticVector_erase

 1 FAILED TEST```

Yggdrasill commented:
> Adding asserts will be fine, I don't know exactly where in the spec it's defined but usually the expectation is the caller will do the right thing and not try erase invalid values.

I've added the asserts and also added test cases for them, which are disabled in release builds. In debug builds it will test for these assertions.

Yggdrasill commented on Source/utils/static_vector.hpp:
```diff
@@ -71,12 +71,26 @@ class StaticVector {
+	void erase(const T *element)
+	{
+		assert(element >= begin() && (element + 1) <= end());
+		erase(element, element + 1);
```
Fixed that now.

ephphatha commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,185 @@
+	}
+
+	size_t erasures = container.size();
+	for (size_t i = 0; i < erasures && container.size() > 0; i++) {
```
nit: you never use `i` in this loop, this could simply be `while (!container.empty())` (also removes the need for the `erasures` constant).

ephphatha reviewed (commented):
Looks good, one more nit and pointing out something you did that I didn't think of but makes sense for the other test case too :)

Yggdrasill commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,185 @@
+
+	size_t erasures = container.size();
+	for (size_t i = 0; i < erasures && container.size() > 0; i++) {
+		size_t idx = RandomIntLessThan(static_cast<int32_t>(container.size()));
```
I agree with this completely. I'll fix that and a few other incorrect uses of `RandomIntBetween()`, where it could trigger assertions when there's no `ASSERT_DEATH()` since it rolls up to size inclusive.

Yggdrasill commented on test/staticvector_test.cpp:
```diff
@@ -0,0 +1,168 @@
+	}
+
+	int32_t from = RandomIntLessThan(static_cast<int32_t>(container.size()));
+	int32_t to = RandomIntBetween(from, static_cast<int32_t>(container.size() - 1), true);
```
I think I've just misunderstood the way that `RandomIntBetween()` works. Thanks for pointing that out! I changed a few other uses of the function given this information too.

ephphatha reviewed (approved):
Nice work, sorry for the back and forth for what is a fairly isolated change 😅 

looks correct to me but it’s getting late in the evening so I’ll hope a second set of eyes can do the merge (or I’ll check again tomorrow)

Yggdrasill commented:
I appreciate the review, don't worry about the back and forth. Thank you.

StephenCWills commented on Source/utils/static_vector.hpp:
```diff
@@ -71,12 +71,26 @@ class StaticVector {
+		T *lastPtr = data_[last - this->begin()].ptr();
+
+		std::move(lastPtr, this->end(), firstPtr);
+		for (const T *it = this->end() - count; it < this->end(); ++it) {
 			std::destroy_at(it);
 		}
```
Does it matter that `std::destroy_at()` is not invoked on the elements that were actually erased, but rather the elements that are left over after `std::move()`?

Yggdrasill replied:
I don't think it's a problem? I believe the current behaviour is also how `std::vector` behaves, which seems to me like what we should copy here. If you're thinking about elements being erased and any existing pointers/indices now being invalid, I think this is also a problem with `std::vector`.

glebm replied:
To me it looks like it is actually called on the left over elements.

Yggdrasill replied:
I've fixed it now. Calls destroy_at() on the elements to be erased, and then does the move operation.

Yggdrasill commented:
So I've researched this more, discussed this some more on Discord, and looked at a libstdc++ implementation of `std::vector`:

```
    _M_erase(iterator __position)
    {
      if (__position + 1 != end())
	_GLIBCXX_MOVE3(__position + 1, end(), __position);
      --this->_M_impl._M_finish;
      _Alloc_traits::destroy(this->_M_impl, this->_M_impl._M_finish);
      _GLIBCXX_ASAN_ANNOTATE_SHRINK(1);
      return __position;
    }
```

This code moves `[__position + 1, end())` to `__position`, then shrinks the size of the vector, destroys the tail element and tells the address sanitiser that the size shrunk by one. I have closely mirrored this in the new commit, and it seems like the correct implementation.

In the new commit, the elements being moved to are overwritten, both with move assignable and copy assignable types. The elements trailing at the end are now garbage, and need to be cleaned up, where `std::destroy` comes in. The elements that shouldn't be erased have already been moved, and `std::move` takes care of cleaning up loose ends like that, so it really shouldn't be a problem. Conceptually, it is somewhat similar to this C code, as far as move assignable types are concerned:

```
int *ptr1 = malloc(sizeof(*ptr1));
int *ptr2 = malloc(sizeof(*ptr2));
free(ptr1);
ptr1 = ptr2;
ptr2 = NULL;
```

glebm commented on Source/utils/static_vector.hpp:
```diff
@@ -87,6 +95,9 @@ class StaticVector {
 
 	void clear()
 	{
+		if (empty()) {
```
This check isn't needed I think

Yggdrasill replied:
I agree, and I have removed it.
